### PR TITLE
fix #19 link delete

### LINKED ISSUE #19: link delete
In user interface, create the possibility to delete links (edges) between nodes.
Description of the successive user actions to delete a link:

- Shift + left mouse button down : select a link
- Delete key : delete link (in UI and in data)

### PR DESCRIPTION
**Pull request by jogal360**

### CODE REVIEW

File: graph-common.js [hunk 1]
```diff
@@ -34,7 +34,7 @@
 
 	damasGraph.prototype.unselectAll = function( ) {
 		this.selection.forEach(function(n){
-			this.getShape(n).classList.remove('selected');
+			graph.getShape(n).classList.remove('selected');
```
Comment (remyla) [inline graph-common.js]: it needs to access the getShape method from within the object but the use of forEach changes the context `this`. Either use the legacy for() loop to keep the context or set the context manually using .bind(this) on function. Getting the global var from within the library is wrong

File: graph-d3.js [hunk 2]
```diff
@@ -125,6 +125,12 @@
 
 	damasGraph.prototype.newEdge = function ( link )
 	{
+		var id_link;
+		if("link_id" in link)
+			id_link = link.link_id;
+		else if("_id" in link)
+			id_link =link._id;
+
```
Comment (remyla) [inline graph-d3.js]: this variable is not used in your code

File: graph-d3.js [hunk 3]
```diff
@@ -145,16 +151,60 @@
...
+		this.restart();
+		return true;
+	}
+
 	damasGraph.prototype.getShape = function( node )
 	{
 		return node.shape;
 	}
+	
+	damasGraph.prototype.link_selected = null;
```
Comment (remyla) [inline graph-d3.js]: I think that link_selected is not used..

File: graph-d3.js [hunk 4]
```diff
@@ -215,7 +265,7 @@
 			if (d3.event.defaultPrevented) return; // click suppressed
 			//assetOverlay(d);
 			if(window['node_pressed']){
-				node_pressed.call(graph.node_lut[d._id], d3.event);
+				node_pressed.call(d, d3.event);
```
Comment (remyla) [inline graph-d3.js]: we like node_pressed to handle genereic nodes and not specific d3 data, so removing the line will break the behavior.. please ask me if you need more info

File: graph-springy.js [hunk 5]
```diff
@@ -51,7 +51,9 @@
 			var springy_target_id = this.springy_lut[l['tgt_id']];
 			var springy_source_node = this.springy_graph.nodes[springy_source_id];
 			var springy_target_node = this.springy_graph.nodes[springy_target_id];
-			this.springy_graph.newEdge(springy_source_node, springy_target_node);
+			var springy_edge = this.springy_graph.newEdge(springy_source_node, springy_target_node);
+			if(l.id && !l._id) l._id = l.id; // backward compatibility
+			this.springy_lut[l._id] = springy_edge.id;
```
Comment (remyla) [inline graph-springy.js]: this code will not work. please have a look at how springy works:
https://github.com/dhotson/springy/blob/master/springy.js#L149
https://github.com/dhotson/springy/blob/master/springy.js#L155

it takes care of 2 separate integers to keep track of edges and nodes

    this.nextNodeId++
    var edge = new Edge(this.nextEdgeId++, source, target, data);

which means that you will have collisions if you use them as index of the same array

File: graph-springy.js [hunk 6]
```diff
@@ -59,15 +61,35 @@
...
 	damasGraph.prototype.removeNode = function( node ){
 
-		if (this.removeNode(node))
+		if (this._removeNode(node))
 		{
+			delete this.node_lut[node.data._id];
 			node.shape.parentNode.removeChild(node.shape);
-			this.springy_graph.removeNode( node );
+			var spr_graph = this.springy_graph;
+			(node.source && node.target) ? spr_graph.removeEdge( node ): spr_graph.removeNode( node );
```
Comment (remyla) [inline graph-springy.js]: the use of node.data._id is confusing

File: graph-springy.js [hunk 7]
```diff
@@ -59,15 +61,35 @@
...
+			search = this.springy_graph.edges;
+		}
+		var idSpringyNodes = this.springy_lut[id];
+		var figure; 
+		search.forEach(function(n){
+			if(n.id == idSpringyNodes)
+				figure = n.shape;
+			//console.log(node.shape);
+		});
+		return figure; //shape
```
Comment (remyla) [inline graph-springy.js]: this will not work, see reason above about springy_lut indexes

File: graph-springy.js [hunk 8]
```diff
@@ -112,6 +134,12 @@
 					{
 						edge.shape.setAttribute('marker-end', 'url(#arrow)' );
 					}
+					edge.shape.addEventListener( 'click', function(e){
+  						if(window['node_pressed']){
+	  						console.log(edge, this);
+	  						node_pressed.call(edge, e);
+   						}
+    				});
```
Comment (remyla) [inline graph-springy.js]: if we give a springy edge object to node_pressed that means that node_pressed will depend on springy types which we would not prefer

**Comment by jogal360** [comment]
New version pull request

**Comment by remyla** [comment]
thank you, well done. I merge your pull request and let you fix the remaining bugs we just spoke about

**Pull request merged**